### PR TITLE
[JENKINS-61900] JCasC support for CredentialsProvider extensions

### PR DESCRIPTION
**Pull request by amuniz**
See [JENKINS-61900](https://issues.jenkins-ci.org/browse/JENKINS-61900).

/cc @jenkinsci/configuration-as-code-plugin-developers

### CODE REVIEW

**Comment by amuniz** [comment]
Looks like I can not add reviewers, so just mentioning here: @timja @jetersen @oleg-nenashev

File: src/test/java/com/cloudbees/plugins/credentials/casc/CredentialsProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        final UsernamePasswordCredentials up = ups.get(0);
+        Assert.assertThat(up.getPassword().getPlainText(), equalTo("password"));
+    }
+
+    @Test
+    @ConfiguredWithCode("CredentialsProviderExtension.yaml")
+    public void export_credentials_provider_extension_credentials() throws Exception {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        ConfigurationAsCode.get().export(out);
+        System.out.println(out.toString());
```
Comment (timja) [inline src/test/java/com/cloudbees/plugins/credentials/casc/CredentialsProviderTest.java:76]: this still needed?

File: src/test/java/com/cloudbees/plugins/credentials/casc/CredentialsProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        );
+        assertThat(dummyCred, hasSize(1));
+        assertThat(dummyCred.get(0).getUsername(), equalTo("user1"));
+
+        // the system provider works fine too
+        List<UsernamePasswordCredentials> ups = CredentialsProvider.lookupCredentials(
+                UsernamePasswordCredentials.class, j.jenkins, ACL.SYSTEM,
+                Collections.singletonList(new HostnameRequirement("api.test.com"))
+        );
+        Assert.assertThat(ups, hasSize(1));
```
Comment (timja) [inline src/test/java/com/cloudbees/plugins/credentials/casc/CredentialsProviderTest.java:66]: i normally static import the Assert classes for improved readability, is there a reason you didn't do this here?

File: src/test/java/com/cloudbees/plugins/credentials/casc/CredentialsProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        final UsernamePasswordCredentials up = ups.get(0);
+        Assert.assertThat(up.getPassword().getPlainText(), equalTo("password"));
```
Comment (timja) [inline src/test/java/com/cloudbees/plugins/credentials/casc/CredentialsProviderTest.java:68]: could we use has plain text? https://github.com/jenkinsci/jenkins-test-harness/blob/master/src/main/java/org/jvnet/hudson/test/JenkinsMatchers.java#L404

```suggestion
       assertThat(up.getPassword(), hasPlainText("password"));
```

File: src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java [hunk 4]
```diff
@@ -77,9 +83,24 @@ public GlobalCredentialsConfiguration instance(Mapping mapping, ConfigurationCon
...
     public Set<Attribute<GlobalCredentialsConfiguration,?>> describe() {
-        return Collections.singleton(new Attribute<GlobalCredentialsConfiguration, SystemCredentialsProvider>("system", SystemCredentialsProvider.class)
-            .getter( t -> SystemCredentialsProvider.getInstance() )
-            .setter( noop() ));
+        HashSet<Attribute<GlobalCredentialsConfiguration, ?>> set = new HashSet<>();
+        Attribute<GlobalCredentialsConfiguration, SystemCredentialsProvider> system = new Attribute<GlobalCredentialsConfiguration, SystemCredentialsProvider>("system", SystemCredentialsProvider.class)
+                .getter(t -> SystemCredentialsProvider.getInstance())
+                .setter(noop());
+        set.add(system);
+        for(CredentialsProvider provider : CredentialsProvider.all()) {
```
Comment (timja) [inline src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java:91]: nit spacing
```suggestion
        for (CredentialsProvider provider : CredentialsProvider.all()) {
```

File: src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java [hunk 5]
```diff
@@ -77,9 +83,24 @@ public GlobalCredentialsConfiguration instance(Mapping mapping, ConfigurationCon
...
+            Symbol s = provider.getClass().getAnnotation(Symbol.class);
+            String name;
+            if (s != null) {
+                name = s.value()[0]; // preferred
+            } else {
+                name = provider.getClass().getSimpleName(); // not so preferred
```
Comment (timja) [inline src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java:97]: should this strip off the `CredentialsProvider` suffix like jcasc does when guessing symbols normally

**Comment by amuniz** [comment]
> the other credential providers I've seen

Can you point me to those please? I'm validating this with an non-public extension, but I'd like to do it with at least another one.

**Comment by timja** [comment]
https://github.com/jenkinsci/aws-secrets-manager-credentials-provider-plugin/blob/master/src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsProvider.java
https://github.com/jenkinsci/azure-keyvault-plugin/blob/master/src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/AzureCredentialsProvider.java

File: src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java (hunk 5 above)
Comment (amuniz) [reply src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java:97]: Do you know of some public API I can use from CasC to get this?

Comment (timja) [reply src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java:97]: https://github.com/jenkinsci/configuration-as-code-plugin/blob/dcd748d6518204f44f7845a7afef80a8b6a601fe/plugin/src/main/java/io/jenkins/plugins/casc/impl/attributes/DescribableAttribute.java#L37

**Comment by amuniz** [comment]
OK. I understand now. So those providers do not create credentials entries in Jenkins, the credentials store is just calling the provider to read credentials from the external system (AWS or Azure).

So, yeah, this is for `CredentialsProvider` extensions which actually store _something_ in Jenkins. In my specific case it's CyberArk, the credentials provider reads from CyberArk but users need to create a credentials entry in Jenkins (just with an ID matching the CyberArk one). This approach has some benefits that are not relevant now.

Thanks for the references!

File: src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java (hunk 5 above)
Comment (amuniz) [reply src/main/java/com/cloudbees/plugins/credentials/casc/CredentialsRootConfigurator.java]: Done. Much better now. Thanks.

**Comment by amuniz** [comment]
Build failed for unrelated reasons: `Remote call on EC2 (aws) - Ubuntu 18.04 LTS (i-096e07c7f640f5d6a) failed`. Re-triggering.

**Comment by jvz** [comment]
Is this ready for review?

**Comment by jetersen** [comment]
Yes, Jenkins CI is just sick 😓

**Pull request merged**